modnet/fit_genetic.py

```diff
@@ -0,0 +1,246 @@
+            mode="auto",
+            baseline=None,
+            restore_best_weights=True,
+        )
+        callbacks = [rlr, es]
+        for w in self.pop:
+            modnet_model = MODNetModel([[['BV_Ea']]], {'BV_Ea':1}, n_feat=w[0], num_neurons=[[int(w[1])],[int(w[1]*w[2])],[int(w[1]*w[2]*w[3])],[int(w[1]*w[2]*w[3]*w[4])]], act=w[5])
+            try:
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 2*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 4*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 8*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                f = mse(modnet_model.predict(X_val),y_val)
+                print('MSE = ', f)
+                self.fitness.append([f, modnet_model, w])
+            except:
+                 pass
+        return self.fitness
+
+
+    def gen_alg(
+        self,
+        X_train: MODData,
+        y_train: pd.DataFrame,
+        X_val: MODData,
+        y_val: pd.DataFrame,
+        size_pop: int,
+        num_epochs: int,
+        prob_mut: float = 0.8
+        )->None:
+
+        """Selects the best individual (the model with the best parameters) for the next generation. The selection is based on a minimisation of the MSE on the validation set.
+
+        Parameters:
+            X_train: Input data of the training set.
+            y_train: Target values of the training set.
+            X_val: Input data of the validation set.
+            y_val: Target values of the validation set.
+            size_pop: Size of the population per generation.
+            num_epochs: Number of generations.
+            prob_mut: Probability the mutation occurs.
+        """
+
+        print('Generation number 0')
+        pop = self.initialization_population(size_pop)
+        fitness = self.function_fitness(pop,  X_train, y_train, X_val, y_val)
+        pop_fitness_sort = np.array(list(sorted(fitness,key=lambda x: x[0])))
+        for j in range(0, num_epochs):
+            print('Generation number ', j+1)
+            length = len(pop_fitness_sort)
+            #select parents
+            parent_1 = pop_fitness_sort[:,2][:length//2]
+            parent_2 = pop_fitness_sort[:,2][length//2:]
+            #crossover
+            child_1 = [self.crossover(parent_1[i], parent_2[i]) for i in range(0, np.min([len(parent_2), len(parent_1)]))]
+            child_2 = [self.crossover(parent_2[i], parent_1[i]) for i in range(0, np.min([len(parent_2), len(parent_1)]))]
+            child_2 = self.mutation(child_2, prob_mut)
```

Comment:
Typically you would define a crossover/mutation ratio, say 0.5, and generate half of your next population from crossover and half from mutation. Not sure what the difference between `child_1` and `child_2` are here.